frontend/src/pages/projects/screens/spawner/spawnerUtils.ts

```diff
@@ -0,0 +1,254 @@
+ * Get all the `imageStream.spec.tags` and filter the ones exists in `imageStream.status.tags`
+ */
+export const getExistingVersionsForImageStream = (
+  imageStream: ImageStreamKind,
+): ImageStreamSpecTagType[] => {
+  const allVersions = imageStream.spec.tags || [];
+  return allVersions.filter((version) => checkVersionExistence(imageStream, version));
+};
+
+/**
+ * Get all the existing tags and check their build status, filter out the available ones
+ */
+export const getAvailableVersionsForImageStream = (
+  imageStream: ImageStreamKind,
+  buildStatuses: BuildStatus[],
+): ImageStreamSpecTagType[] => {
+  const existingVersions = getExistingVersionsForImageStream(imageStream);
+  return existingVersions.filter((version) =>
+    checkTagBuildValid(buildStatuses, imageStream, version),
+  );
+};
+
+/**
+ * Get the recommended tag or the default tag
+ * if no recommended or default, return the latest version
+ */
+export const getDefaultVersionForImageStream = (
+  imageStream: ImageStreamKind,
+  buildStatuses: BuildStatus[],
+): ImageStreamSpecTagType | undefined => {
+  const availableVersions = getAvailableVersionsForImageStream(imageStream, buildStatuses);
+  if (availableVersions.length === 0) {
+    return undefined;
+  }
+
+  const sortedVersions = [...availableVersions].sort(compareTagVersions);
+
+  const defaultVersion = sortedVersions.find(
+    (version) =>
+      version.annotations?.['opendatahub.io/notebook-image-recommended'] ||
+      version.annotations?.['opendatahub.io/default-image'],
+  );
+  if (defaultVersion) {
+    return defaultVersion;
+  }
+
+  // Return the most recent version
+  return sortedVersions[0];
+};
+
+/******************* Deployment Size utils *******************/
+export const getSizeDescription = (size: NotebookSize): string =>
+  `Limits: ${size.resources.limits?.cpu || '??'} CPU, ` +
+  `${size.resources.limits?.memory || '??'} Memory ` +
+  `Requests: ${size.resources.requests?.cpu || '??'} CPU, ` +
+  `${size.resources.requests?.memory || '??'} Memory`;
+
+/******************* Checking utils *******************/
+/**
+ * Check if there is 1 or more versions available for an image stream
+ */
+export const checkImageStreamAvailability = (
+  imageStream: ImageStreamKind,
+  buildStatuses: BuildStatus[],
+): boolean => {
+  const tags = getAvailableVersionsForImageStream(imageStream, buildStatuses);
+  return tags.length !== 0;
+};
+
+export const checkTagBuildValid = (
+  buildStatuses: BuildStatus[],
+  imageStream: ImageStreamKind,
+  imageVersion: ImageStreamSpecTagType,
+): boolean => {
+  const imageStreamVersion = `${imageStream.metadata.name}:${imageVersion.name}`;
+  const build = buildStatuses.find(
+    (buildStatus) => buildStatus.imageStreamVersion === imageStreamVersion,
+  );
+  if (!build) {
+    return true;
+  }
+  return !PENDING_PHASES.includes(build.status) && !FAILED_PHASES.includes(build.status);
+};
+
+/**
+ * Check whether `imageStream.spec.tags[i]` exists in status.tags
```

Comment:
Nit, it is a bit odd to explain how something is done in code in a block comment telling you about the function. Functionality can always change -- comments become stale and often are left around.